3-reveal/migrations/6-MDA/Zambia-2020/deploy/mda_dispense_tasks.psql

```diff
@@ -0,0 +1,221 @@
+        COALESCE(COUNT(latest_events.id) FILTER (WHERE latest_events.form_data->>'availableForTreatment' IN ('Yes')), 0) AS availableForTreatment,
+        COALESCE(SUM(COALESCE(latest_events.form_data->>'nPzqDistributedQuantity', '0')::DECIMAL), 0) AS nPzqDistributedQuantity
+    FROM tasks
+    LEFT JOIN LATERAL
+    	(
+    		SELECT
+    			events.*
+    		FROM events
+    		WHERE
+    			events.task_id = tasks.identifier
+    			AND events.event_type = 'mda_dispense'
+    		ORDER BY events.server_version DESC
+    		LIMIT 1
+    	) AS latest_events
+    	ON TRUE
+    WHERE
+    	tasks.code = 'MDA_Dispense'
+        AND latest_events.id IS NOT NULL
+    GROUP BY tasks.plan_identifier, tasks.group_identifier
+)
+SELECT
+    metrics.*,
+    jurisdiction_hierarchy.jurisdiction_path || jurisdiction_hierarchy.jurisdiction_id AS full_jurisdiction_path
+FROM metrics
+LEFT JOIN jurisdictions_materialized_view AS jurisdiction_hierarchy
+    ON jurisdiction_hierarchy.jurisdiction_id = metrics.jurisdiction_id;
+
+CREATE INDEX IF NOT EXISTS ntd_dispense_task_metrics_plan_id_jurisdiction_id_idx ON ntd_dispense_task_metrics (plan_id, jurisdiction_id);
+CREATE INDEX IF NOT EXISTS ntd_dispense_task_metrics_plan_id_full_jurisdiction_path_idx ON ntd_dispense_task_metrics USING GIN(full_jurisdiction_path);
+
+-- Compute aggregate jurisdiction metrics
+CREATE MATERIALIZED VIEW ntd_dispense_plan_metrics AS
+SELECT
+    plan_hierarchy.plan_id,
+    plan_hierarchy.jurisdiction_id,
+    BOOL_AND(plan_hierarchy.is_focus_area) AS irs_focus_area,
+    SUM(ntd_dispense_task_metrics.nPzqDistribute) AS nPzqDistribute,
+    SUM(ntd_dispense_task_metrics.availableForTreatment) AS availableForTreatment,
+    SUM(ntd_dispense_task_metrics.nPzqDistributedQuantity) AS nPzqDistributedQuantity
+FROM plan_jurisdictions_materialized_view AS plan_hierarchy
+LEFT JOIN ntd_dispense_task_metrics
+    ON ntd_dispense_task_metrics.plan_id = plan_hierarchy.plan_id
+    AND plan_hierarchy.jurisdiction_id = ANY(ntd_dispense_task_metrics.full_jurisdiction_path)
+GROUP BY plan_hierarchy.plan_id, plan_hierarchy.jurisdiction_id;
+
+CREATE INDEX IF NOT EXISTS ntd_dispense_plan_metrics_plan_id_jurisdiction_id_idx ON ntd_dispense_plan_metrics (plan_id, jurisdiction_id);
+
+--
+--
+--
+
+CREATE MATERIALIZED VIEW ntd_visit_structure_task_metrics AS
+WITH metrics AS
+(
+    SELECT
+        tasks.plan_identifier AS plan_id,
+        tasks.group_identifier AS jurisdiction_id,
+        COUNT(tasks.identifier) AS structures_visited,
+        COALESCE(SUM(COALESCE(clients.attributes->>'nsac', '0')::DECIMAL), 0) AS sacregistered
+    FROM tasks
+    LEFT JOIN events
+        ON tasks.identifier = events.task_id
+        AND events.event_type IN ('Family Registration', 'Family_Registration')
+    LEFT JOIN clients
+        ON events.base_entity_id = clients.baseentityid
+    WHERE
+        tasks.code IN ('Structure_Visited', 'Structure Visited')
+        AND tasks.status IN ('Completed')
```

Comment:
why do we not restrict on `tasks.status` in the above query? Are they not used for MDA or we don't care about restricting on them?

Reply:
@moshthepitt actually worked on the metrics calculation itself, so I was trying not to touch it because often these kind of changes matter in that we don't match the app exactly.  I suspect it'd be okay, but would require thought (and maybe a PM?).

I think a general challenge I have with these business-logic views is trying to optimize at the overall data structure level vs optimizing the particular business logic chosen.  I worry when the overall performance of our system is tied so closely to the particular aggregation choices a client makes, because they change those a lot.

That's vague, I guess, I'm just saying maybe in the future we could think about first creating matviews that link all the (latest) events and clients and structures and tasks together as one more-expensive step, then aggregating more cheaply in different ways on top of that kind of matview vs doing that linking in every metric we have.  90% of the reporting I've seen is basically that linking and then sums.

Reply:
> I guess, I'm just saying maybe in the future we could think about first creating matviews that link all the (latest) events and clients and structures and tasks together as one more-expensive step, then aggregating more cheaply in different ways on top of that kind of matview vs doing that linking in every metric we have. 90% of the reporting I've seen is basically that linking and then sums

I think this is interesting and we should explore it further.  Now that we have a large number of different Reveal reports we can look at it overall and identify how to streamline this and make it generally better for everyone.